extensions/protobuf/src/main/java/io/deephaven/protobuf/FieldPath.java

```diff
@@ -0,0 +1,124 @@
+/**
+ * Copyright (c) 2016-2023 Deephaven Data Labs and Patent Pending
+ */
+package io.deephaven.protobuf;
+
+import com.google.protobuf.Descriptors.FieldDescriptor;
+import io.deephaven.annotations.SimpleStyle;
+import io.deephaven.functions.BooleanFunction;
+import org.immutables.value.Value.Immutable;
+import org.immutables.value.Value.Lazy;
+import org.immutables.value.Value.Parameter;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static io.deephaven.functions.BooleanFunction.map;
+
+/**
+ * The {@link #path()} to a {@link com.google.protobuf.Descriptors.Descriptor Descriptor's} field.
+ *
+ * <p>
+ * {@link FieldDescriptor} objects are not equal across {@link com.google.protobuf.DynamicMessage dynamic messages}.
+ * Helpers {@link #numberPath()} and {@link #namePath()} may be used in lieu of direct equivalence depending on context.
+ */
+@Immutable
+@SimpleStyle
+public abstract class FieldPath {
+
+    public static FieldPath empty() {
+        return of(List.of());
+    }
+
+    public static FieldPath of(FieldDescriptor... descriptors) {
+        return of(Arrays.asList(descriptors));
+    }
+
+    public static FieldPath of(List<FieldDescriptor> descriptors) {
+        return ImmutableFieldPath.of(descriptors);
+    }
+
+    /**
+     * Creates a function that returns {@code true} if {@code namePath} starts with the {@link FieldPath#namePath()
+     * field path's name path}.
+     *
+     * @param namePath the name path
+     * @return the boolean function
+     * @see FieldPath#otherStartsWithThis(List)
+     */
+    public static BooleanFunction<FieldPath> namePathStartsWithFieldPath(List<String> namePath) {
```

Comment:
IMO, anytime DH has the option, we should return the more rich type (so, it can work natively w/ the additional tooliing. Conversely, whenever possible, we should take the more general parameter type.